### PR TITLE
ENH: CloughTocher2DInterpolator multiple times with different values

### PR DESCRIPTION
**Pull request by hovavalon**
#### Reference issue
#13306 

#### What does this implement/fix?
The PR adds an ability to change interpolation values in an existing `CloughTocher2DInterpolator` object, while also saving the barycentric coordinates of interpolation points.
In a use case in which only the values at the given points change, this saves most of the function's runtime.
The PR includes basic tests.

#### Additional information
This is my first PR to `scipy`, and I am not 100% sure how to document the API, help and guidance would be appreciated.

### CODE REVIEW

**Comment by hovavalon** [comment]
Hi, I am having trouble understanding the outputs of the failed CI tests.
When I compile the code locally with meson it works, but the CI fails with the following error without additional explanation:
`meson.build:1:0: ERROR: Cython compiler 'cython' cannot compile programs`
How should I approach this?

**Comment by ev-br** [comment]
This error you can ignore, it's not related to your changes and is being worked on separately.

**Comment by ev-br** [comment]
Have to admit, I don't like this at its present form. The reasons are mirroring https://github.com/scipy/scipy/pull/10860#issuecomment-1118610127: it adds essentially a global state to objects and the result is hard to reason about and hard to maintain.

Maybe there is a way to rework this similar to gh-17230 : add necessary hooks for a downstream subclass to override.

**Comment by hovavalon** [comment]
Ok, I see your point.
I have added `_set_xi` and `_set_values` functions to `NDInterpolatorBase` and implemented them for `CloughTocher2DInterpolator`.
Also, I have updated the tests accordingly.
What do you think?

**Comment by hovavalon** [comment]
@ev-br I would like to know your opinion on the following matter:
The function `_set_xi` I implemented in `CloughTocher2DInterpolator` is also relevant to `LinearNDInterpolator`, but not to `NearestNDInterpolator` which also subclasses `NDInterpolatorBase`.
How should I approach the code reuse of `_set_xi` in this case?

Also, the function `qhull._barycentric_coordinates` is called in `qhull._find_simplex` for every point, but then even though `_clough_tocher_2d_single` obtains the output, it overrides it.
In the case I am working on removing it saved 10% of `_do_evaluate`'s runtime in the `CloughTocher2DInterpolator`, Is it OK if I add the commit removing it to this PR?

**Comment by ev-br** [comment]
> The function _set_xi I implemented in CloughTocher2DInterpolator is also relevant to LinearNDInterpolator, but not to > NearestNDInterpolator which also subclasses NDInterpolatorBase.
> How should I approach the code reuse of _set_xi in this case?

I'd just override it in `NearestNDInterpolator` with an empty implementation + add a code comment to explain it.

> Also, the function qhull._barycentric_coordinates is called in qhull._find_simplex for every point, but then even though _clough_tocher_2d_single obtains the output, it overrides it.
In the case I am working on removing it saved 10% of _do_evaluate's runtime in the CloughTocher2DInterpolator, Is it OK if I add the commit removing it to this PR?

Certainly, if it's not needed, it's best removed. That's an _if_ though, this code has not been looked at for quite a while, it'd be great if you could take a close look.

One other thing which would be useful is to do an asv benchmark (similar to what was done for the RGI) to double-check there are no slowdowns.

**Comment by hovavalon** [comment]
I have updated the PR such that now the barycentric coordinates are calculated in `NDInterpolatorBase._set_xi`, which isn't called in `NDInterpolatorBase`.

About the call to `qhull._barycentric_coordinates`, I think it might be necessary after all and will look into it at some point, but I think I it will not get into this PR.

**Comment by ev-br** [comment]
Note that CI errors are real and need to be fixed.

File: scipy/interpolate/interpnd.pyx [hunk 1]
```diff
@@ -55,8 +56,8 @@ class NDInterpolatorBase:
 
     """
 
-    def __init__(self, points, values, fill_value=np.nan, ndim=None,
-                 rescale=False, need_contiguous=True, need_values=True):
+    def __init__(self, points, values=None, fill_value=np.nan, ndim=None,
+                 rescale=False, need_contiguous=True, need_values=True, xi=None):
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:60]: Cannot do this, the signature must be preserved. `values` need to stay a required argument, and `xi` cannot appear here. You create an interpolator with given `values` and call it with given new coordinates `xi`.

Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:60]: I am quoting from my original PR:
```
The PR adds an ability to change interpolation values in an existing CloughTocher2DInterpolator object, while also saving the barycentric coordinates of interpolation points.
```
Is there a better way to do this?
This change in the API doesn't affect existing code at all, and without changing it this way, or at least with the addition of `xi=None`, there is no way to save the barycentric coordinates between runs without maintining a problematic global state to the object.

**Comment by ev-br** [comment]
My suggestion is that users maintain a subclass: https://github.com/scipy/scipy/pull/18376#issuecomment-1529517547

**Comment by hovavalon** [comment]
@ev-br Ok, I misunderstood your first comments, sorry for that :(
I have removed the changes to the existing API, and kept the `_set_values` and `_set_xi` functions for possible usage in subclasses.
What do you think?

File: scipy/interpolate/_ndgriddata.py [hunk 2]
```diff
@@ -92,12 +92,17 @@ class NearestNDInterpolator(NDInterpolatorBase):
     def __init__(self, x, y, rescale=False, tree_options=None):
         NDInterpolatorBase.__init__(self, x, y, rescale=rescale,
                                     need_contiguous=False,
-                                    need_values=False)
+                                    need_values=True)
```
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:95]: Can we remove this too? This is not needed by vanilla scipy interpolators; a downstream subclass either knows it needs it or can add this argument ISTM.

File: scipy/interpolate/_ndgriddata.py [hunk 3]
```diff
@@ -92,12 +92,17 @@ class NearestNDInterpolator(NDInterpolatorBase):
...
         if tree_options is None:
             tree_options = dict()
         self.tree = cKDTree(self.points, **tree_options)
         self.values = np.asarray(y)
 
+    def _calculate_triangulation(self):
+        # This function overrides _calculate_triangulation in NDInterpolatorBase,
+        # since triangulation is not needed for NearestNDInterpolator.
+        pass
```
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:104]: So this is a subroutine which attaches a bunch of things to self. What is it supposed to do modify and can it just return whatever it needs to return. Then attaching the values is best done at the call site.

File: scipy/interpolate/_ndgriddata.py [hunk 4]
```diff
@@ -110,6 +115,8 @@ def __call__(self, *args):
             or x1 can be array-like of float with shape ``(..., ndim)``
 
         """
+        # This function doesn't use NDInterpolatorBase._set_xi since it performs
+        # unrequired calculations.
```
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:119]: This seems to be a leftover.

File: scipy/interpolate/interpnd.pyx [hunk 5]
```diff
@@ -73,36 +73,15 @@ class NDInterpolatorBase:
         else:
             self.tri = None
 
+        self.ndim = ndim
         points = _ndim_coords_from_arrays(points)
-        values = np.asarray(values)
 
-        _check_init_shape(points, values, ndim=ndim)
+        self.need_contiguous = need_contiguous
+        self._original_fill_value = fill_value
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:80]: This version attaches a bunch of attributes to the instance. Let's avoid this: let's instead add hooks for subclasses to override. Subclasses are free to store whatever they wish on an instance, but the scipy version should not unless there is a _very_ compelling reason.

File: scipy/interpolate/interpnd.pyx [hunk 6]
```diff
@@ -113,6 +92,46 @@ class NDInterpolatorBase:
...
+            self._set_values(values)
+        else:
+            self.values = None
+        
+        self._xi = None
+        self._xi_simplices = None
+        self._xi_barycentric_coordinates = None
+
+    def _calculate_triangulation(self):
+        self.tri = qhull.Delaunay(self.points)
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:108]: `self.tri = self._calculate_triangulation(self.points)`.

Also maybe this can be added to LinearNDInterpolator and CTInterpolator only, and not to the Base. Yes, this breaks LSP a little but so what.

**Review by ev-br — CHANGES_REQUESTED** [review]
This is getting closer, but there's still work to do. Ideally, we don't attach a bunch of attributes to instances.

File: scipy/interpolate/interpnd.pyx (hunk 6 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:108]: Done.

File: scipy/interpolate/interpnd.pyx (hunk 5 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:80]: Will fix.

File: scipy/interpolate/_ndgriddata.py (hunk 4 above)
Comment (hovavalon) [reply scipy/interpolate/_ndgriddata.py:119]: Actually, it isn't. I wanted to explain why this subclass has its own implementation for this function.

File: scipy/interpolate/_ndgriddata.py (hunk 3 above)
Comment (hovavalon) [reply scipy/interpolate/_ndgriddata.py:104]: Fixed, according to the other comment about `_calculate_triangulation`.

File: scipy/interpolate/_ndgriddata.py (hunk 2 above)
Comment (hovavalon) [reply scipy/interpolate/_ndgriddata.py:95]: I think it can be removed, but won't it break existing usages of the current subclasses which use this attribute?
Would you like me to deprecate it somehow?
If so, can you direct me at an example of how deprecation works in scipy?

File: scipy/interpolate/interpnd.pyx (hunk 5 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:80]: I chose to pass these values to the `_set_values` functions, is it OK?

**Comment by hovavalon** [comment]
What do you think about the current form?
I have also rebased on the main branch and solved conflicts.

File: scipy/interpolate/_ndgriddata.py (hunk 2 above)
Comment (ev-br) [reply scipy/interpolate/_ndgriddata.py:95]: Ah, yes, the parameter was there, no need to deprecate it!

File: scipy/interpolate/_ndgriddata.py (hunk 4 above)
Comment (ev-br) [reply scipy/interpolate/_ndgriddata.py:114]: Took me a while to figure out what is referred to as this function. But this is minor, let's sort out the big picture first (cf my other comment about the `__call__).

File: scipy/interpolate/interpnd.pyx [hunk 7]
```diff
@@ -909,37 +965,56 @@ class CloughTocher2DInterpolator(NDInterpolatorBase):
...
+            xi = self._xi
         return self._do_evaluate(xi, 1.0)
 
-    def _evaluate_complex(self, xi):
+    def _evaluate_complex(self, xi=None):
+        if xi is None:
+            xi = self._xi
         return self._do_evaluate(xi, 1.0j)
-
+
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:1000]: Nit: trailing whitespace is typically frowned upon.

File: scipy/interpolate/interpnd.pyx [hunk 8]
```diff
@@ -138,21 +191,24 @@ class NDInterpolatorBase:
...
-        xi = self._check_call_shape(xi)
-        shape = xi.shape
-        xi = xi.reshape(-1, shape[-1])
-        xi = np.ascontiguousarray(xi, dtype=np.double)
+        if len(args) > 0:
+            self._set_xi(*args)
+        elif self._xi is None:
+            raise ValueError("Interpolation was called without required points")
+        if self.values is None:
+            raise ValueError("Values at interpolation points were not set")
```
[the remark this answers is not in the document]
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:203]: Actually the intention was to avoid the operation of finding the barycentric coordinates of points from `xi` in the triangulation.
In order to do so, the cython `evaluate` function should not perform this operation if it was already done, and I am not quite sure how to enable this behaviour for subclasses which won't need to reimplement `do_evaluate`, without saving some kind of state.
What do you suggest?

[the remark this answers is not in the document]
Comment (ev-br) [reply scipy/interpolate/interpnd.pyx:203]: My suggestion is to wrap a call to `_find_simplicies` (or its analog), `_set_xi` and other things into calls that return the necessary info. Then the `CloughTocherInterpolator` class does not store anything on an instance. Your subclass 
- overrides `__init__` to store stuff you want to store on the instance,
- overrides `_find_simplices` et all to look up instance variables.

This way the library version remains simple, and your subclass you can customize however you see fit.

File: scipy/interpolate/interpnd.pyx (hunk 7 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx]: Removed

File: scipy/interpolate/interpnd.pyx (hunk 8 above)
[the remark this answers is not in the document]
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx]: I have changed the functions accordingly.
Now the original signatures remain, and overriding `_find_simplices` should enable my requested fuctionality.

File: scipy/interpolate/_ndgriddata.py [hunk 9]
```diff
@@ -92,7 +92,7 @@ class NearestNDInterpolator(NDInterpolatorBase):
     def __init__(self, x, y, rescale=False, tree_options=None):
         NDInterpolatorBase.__init__(self, x, y, rescale=rescale,
                                     need_contiguous=False,
-                                    need_values=False)
+                                    need_values=True)
```
Comment (ev-br) [inline scipy/interpolate/_ndgriddata.py:95]: What is the effect of the change of the default value and is this change really needed?

File: scipy/interpolate/interpnd.pyx [hunk 10]
```diff
@@ -73,36 +73,12 @@ class NDInterpolatorBase:
         else:
             self.tri = None
 
+        self.ndim = ndim
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:76]: A new attribute, a leftover from previous iterations?

File: scipy/interpolate/interpnd.pyx [hunk 11]
```diff
@@ -846,6 +880,9 @@ class CloughTocher2DInterpolator(NDInterpolatorBase):
     gradients necessary for this are estimated using the global
     algorithm described in [Nielson83]_ and [Renka84]_.
 
+    Since most of the algorithm depeneds only on the triangulation of
+    the input
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:884]: incomplete sensense

File: scipy/interpolate/interpnd.pyx [hunk 12]
```diff
@@ -909,12 +946,28 @@ class CloughTocher2DInterpolator(NDInterpolatorBase):
 
     def __init__(self, points, values, fill_value=np.nan,
                  tol=1e-6, maxiter=400, rescale=False):
+        self.tol = tol
+        self.maxiter = maxiter
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:950]: New attributes, should be removed or motivated.

File: scipy/interpolate/interpnd.pyx [hunk 13]
```diff
@@ -832,6 +864,8 @@ class CloughTocher2DInterpolator(NDInterpolatorBase):
         Rescale points to unit cube before performing interpolation.
         This is useful if some of the input dimensions have
         incommensurable units and differ by many orders of magnitude.
+    xi : tuple of ndarrays, optional
+        The coordinates of the points to be interpolated.
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:868]: This is not a parameter of the `__init__` function.

File: scipy/interpolate/interpnd.pyx [hunk 14]
```diff
@@ -929,17 +982,18 @@ class CloughTocher2DInterpolator(NDInterpolatorBase):
         cdef const double_or_complex[:,:,:] grad = self.grad
         cdef double_or_complex[:,::1] out
         cdef const int[:,::1] simplices = self.tri.simplices
-        cdef double c[NPY_MAXDIMS]
+        cdef double[:,:] c
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:985]: I wonder if the original was a perf optimization. At any rate, we'd need a benchmark to make sure there is no performance regression (C array -> cython memoryview, python functions in `__call__` etc).

File: scipy/interpolate/interpnd.pyx [hunk 15]
```diff
@@ -929,17 +982,18 @@ class CloughTocher2DInterpolator(NDInterpolatorBase):
...
         cdef double_or_complex df[2*NPY_MAXDIMS+2]
         cdef double_or_complex w
         cdef double_or_complex fill_value
-        cdef int i, j, k, ndim, isimplex, start, nvalues
+        cdef int i, j, k, ndim, isimplex, nvalues
         cdef qhull.DelaunayInfo_t info
-        cdef double eps, eps_broad
+        cdef int[:] isimplices
+
+        isimplices,c = self._find_simplicies(xi)
```
Comment (ev-br) [inline scipy/interpolate/interpnd.pyx:994]: This was in a nogil section, any specific reason for moving it out

**Review by ev-br — CHANGES_REQUESTED** [review]
This is getting close!

I've left several small-ish comments; two big-picture asks are:
- can you verify that a subclass does what you want it to?
- we'd need a benchmark to check that there is no big perf regression (might want to take a look at what RGI PR did).

File: scipy/interpolate/_ndgriddata.py (hunk 9 above)
Comment (hovavalon) [reply scipy/interpolate/_ndgriddata.py:95]: This is a residue of a discarded change, replacing with original value.

File: scipy/interpolate/interpnd.pyx (hunk 10 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:76]: Yes, removing it.

Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:76]: Yes, removed.

File: scipy/interpolate/interpnd.pyx (hunk 11 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:884]: Removed.

File: scipy/interpolate/interpnd.pyx (hunk 12 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:950]: These two attributes are used in the `_set_values` function of the `CloughTocher` class, which is an override for the one of `NDInterpolatorBase`, hence they are saved here and used afterwards.

File: scipy/interpolate/interpnd.pyx (hunk 13 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:868]: Removed.

File: scipy/interpolate/interpnd.pyx (hunk 15 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:994]: Every part of this function which was inside a `nogil` clause is still inside one, there is a `nogil` around the main loop of the function.

File: scipy/interpolate/interpnd.pyx (hunk 14 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx:985]: Since all the changed code is used by `griddata` with different methods, and `griddata` already has a benchmark, is it necessary to add another one?
I am adding a benchmark for the subclass anyway, same as in RGI.

**Comment by hovavalon** [comment]
I have fixed the code according to your comments, and added a benchmark test with an example subclass.
As I noted, `griddata` is already benchmarked, so I don't think it is necessary to benchmark the old code.
I had problems running the benchmark, getting the following error:
```
Fatal error in launcher: Unable to create process using '"D:\bld\asv_1636402977906\_h_env\python.exe"  "C:\Users\USER\anaconda3\envs\scipy-dev\Scripts\asv.exe" ': The system cannot find the file specified.
```
If you have any clue on how to solve this issue I would be greateful.
Thanks!

**Comment by ev-br** [comment]
Don't know what the error is, sorry. Something about asv (which is the benchmark runner), so it's something about how you installed scipy. As a data point, I just run a benchmarks for a first time in several years and it basically worked (more below). My dev setup is a `$ mamba env create -f environment.yml` on linux, so bog-standard.

There is indeed some issue with `bench --compare`. However just running a single benchmark works and I can switch branches manually.   It looks like this PR is some 1.7 times slower than main for larger input sizes. This needs looking into.


Running `$ python dev.py bench -t GridData` I get:

```
On a PR (rebased on main):
==========================

$ python dev.py bench -t GridData
· No `environment_type` specified in asv.conf.json. This will be required in the future.
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_home_br_mambaforge_envs_scipy-dev_bin_python
[ 50.00%] ··· Running (interpolate.GridData.time_evaluation--).
[100.00%] ··· interpolate.GridData.time_evaluation                                    ok
[100.00%] ··· ========= ============ ============= =============
              --                         method                 
              --------- ----------------------------------------
               n_grids    nearest        linear        cubic    
              ========= ============ ============= =============
                 10j      275±3μs     6.21±0.05ms   7.24±0.08ms 
                 100j    3.82±0.2ms   8.62±0.08ms    11.3±0.5ms 
                1000j     320±5ms       181±2ms       331±1ms   
              ========= ============ ============= =============



On main ('1.12.0.dev0+1164.5f9e871') :
================================

$ python dev.py bench -t GridData
💻  ninja -C /home/br/repos/scipy/scipy/build -j4
ninja: Entering directory `/home/br/repos/scipy/scipy/build'
[7/7] Linking target scipy/interpolate/interpnd.cpython-310-x86_64-linux-gnu.so
Build OK
💻  meson install -C build --only-changed
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /home/br/repos/scipy/scipy/build-install/lib/python3.10/site-packages
Running benchmarks for Scipy version 1.12.0.dev0+1164.5f9e871 at /home/br/repos/scipy/scipy/build-install/lib/python3.10/site-packages/scipy
· No `environment_type` specified in asv.conf.json. This will be required in the future.
· Discovering benchmarks
· Running 1 total benchmarks (1 commits * 1 environments * 1 benchmarks)
[  0.00%] ·· Benchmarking existing-py_home_br_mambaforge_envs_scipy-dev_bin_python
[ 50.00%] ··· Running (interpolate.GridData.time_evaluation--).
[100.00%] ··· interpolate.GridData.time_evaluation                                    ok
[100.00%] ··· ========= ============ ============ ============
              --                        method                
              --------- --------------------------------------
               n_grids    nearest       linear       cubic    
              ========= ============ ============ ============
                 10j      289±10μs    6.69±0.3ms   7.75±0.6ms 
                 100j    3.88±0.2ms   7.42±0.2ms   9.02±0.2ms 
                1000j     339±3ms      58.3±1ms     191±2ms   
              ========= ============ ============ ============
```

**Comment by hovavalon** [comment]
I will look into it an update when I find the reasons.
For now I have pushed a rebased version of the branch, you may ignore it.

**Comment by hovavalon** [comment]
Since I can't run the benchmarks on my computer for now, I have created a small benchmark of my own, and the differences you display do not show in it, even though it is based completely on the `GridData` benchmark.

The code is as follows:
```
from scipy import interpolate
import numpy as np


import cProfile
for method in ['nearest', 'linear', 'cubic']:
    s = f"""
import numpy as np
from scipy import interpolate
class GridData:

    def __init__(self, n_grids, method):
        self.func = lambda x, y: x*(1-x)*np.cos(4*np.pi*x) * np.sin(4*np.pi*y**2)**2
        self.grid_x, self.grid_y = np.mgrid[0:1:n_grids, 0:1:n_grids]
        self.points = np.random.rand(1000, 2)
        self.values = self.func(self.points[:, 0], self.points[:, 1])
        self.method = method

    def time_evaluation(self):
        interpolate.griddata(self.points, self.values, (self.grid_x, self.grid_y), method=self.method)
grid_data = GridData(1000, '{method}')
"""
    import timeit
    data = timeit.repeat("grid_data.time_evaluation()", s, number=100, repeat=15)
    print(f"# {method}:")
    print(data)
    print(f"# Mean: {np.mean(data)}")
    print(f"# STD: {np.std(data)}")
```


And I obtain the following results:

```
#Branch:

# nearest:
[0.01634670002385974, 0.016986999893561006, 0.01597739988937974, 0.015783799812197685, 0.014918799977749586, 0.015606899978592992, 0.01522199995815754, 0.015267099952325225, 0.01469009998254478, 0.016034400090575218, 0.015890099806711078, 0.017759300069883466, 0.016209400026127696, 0.015358799835667014, 0.015316699864342809]
# Mean: 0.01582456661077837
# STD: 0.0007699367436863699
# linear:
[3.3771730000153184, 3.081603200174868, 3.154085800051689, 3.052978300023824, 3.125168399885297, 3.0178121998906136, 3.0375663999002427, 3.0500803000759333, 3.041135599836707, 3.074869600124657, 3.0457837998401374, 2.956153300125152, 3.0647458999883384, 2.901043600169942, 5.2550554999616]
# Mean: 3.215683660004288
# STD: 0.5542656244427543
# cubic:
[3.218328900169581, 3.1084858998656273, 3.276013000169769, 3.171105199959129, 3.255917799891904, 3.2812592999543995, 3.093714299844578, 3.1348842000588775, 3.3422193999867886, 3.104086200008169, 3.206670899875462, 3.169537999900058, 3.154295900138095, 3.2070639999583364, 3.1233502000104636]
# Mean: 3.189795546652749
# STD: 0.07185193069796424

#Main:
# nearest:
[0.016603800002485514, 0.016050299862399697, 0.015192899852991104, 0.015695499954745173, 0.015167100122198462, 0.01509879995137453, 0.014981599990278482, 0.02002110006287694, 0.016280499985441566, 0.014667100040242076, 0.014549100073054433, 0.014966800110414624, 0.015341500053182244, 0.015880800085142255, 0.016552599845454097]
# Mean: 0.015803299999485414
# STD: 0.0012916204980256561
# linear:
[8.991146600106731, 6.971040299860761, 3.647712399950251, 3.721146099967882, 2.8603197000920773, 3.0786919000092894, 3.758976799901575, 2.8815309000201523, 3.017121700104326, 2.9204843998886645, 2.8852842000778764, 2.95154959987849, 2.880782499909401, 3.0072578999679536, 2.8994682000484318]
# Mean: 3.7648342133189243
# STD: 1.7222134976627812
# cubic:
[3.002680500037968, 3.09849979984574, 2.945954800117761, 2.956872000126168, 3.4364565999712795, 3.616613200167194, 3.1236976999789476, 2.999945000046864, 2.909005500143394, 3.0166422000620514, 2.9343109000474215, 2.9666826999746263, 3.161001200089231, 2.9111989999655634, 3.0782538999337703]
# Mean: 3.077187666700532
# STD: 0.19408612016837506
```

The changes are all within the standard deviation.
Would you mind running this code on your system and uploading your results here?
Thanks!

**Comment by hovavalon** [comment]
Ok, so what else is needed for the PR to be approved?

File: scipy/interpolate/interpnd.pyx (hunk 12 above)
Comment (ev-br) [reply scipy/interpolate/interpnd.pyx:944]: We should not add public attributes unless really necessary.
Consider either not attaching them to the instance (like it is on main), or adding a leading underscore to mark them as private, like `self._tol`

File: scipy/interpolate/_ndgriddata.py (hunk 4 above)
Comment (ev-br) [reply scipy/interpolate/_ndgriddata.py:114]: Please consider rephrasing to clarify what is "this function".

**Review by ev-br — COMMENTED** [review]
OK, this is getting very very close. Two small comments, and then
- if someone has an idea of what's going on with asv benchmarks, that would be helpful
- can you verify that this PR improves the OP use case?

**Comment by hovavalon** [comment]
Here is a working example for subclassing:
```
from scipy import interpolate
import numpy as np



shared_setup = f"""
from scipy import interpolate
import numpy as np
n_grids = 1000

func = lambda x, y: x*(1-x)*np.cos(4*np.pi*x) * np.sin(4*np.pi*y**2)**2
grid_x, grid_y = np.mgrid[0:1:n_grids, 0:1:n_grids]
points = np.random.rand(1000, 2)
values = [func(points[:, 0], points[:, 1]) + t for t in np.arange(-1,1.1,0.1)]
"""

oldclass = """
for val in values:
    interpolate.CloughTocher2DInterpolator(points, val)(grid_x, grid_y)
"""

newclass_setup = """
class CloughTocherInterpolatorValues(interpolate.CloughTocher2DInterpolator):
    def __init__(self, points, xi, tol=1e-6, maxiter=400, **kwargs):
        interpolate.CloughTocher2DInterpolator.__init__(self, points, None, tol=tol, maxiter=maxiter)
        self.xi = None
        self._preprocess_xi(*xi)
        self.simplices, self.c = interpolate.CloughTocher2DInterpolator._find_simplicies(self, self.xi)

    def _preprocess_xi(self, *args):
        if self.xi is None:
            self.xi, self.interpolation_points_shape = interpolate.CloughTocher2DInterpolator._preprocess_xi(self, *args)
        return self.xi, self.interpolation_points_shape
    
    def _find_simplicies(self, xi):
        return self.simplices, self.c

    def __call__(self, values):
        self._set_values(values)
        return super().__call__(self.xi)
"""

newclass_code = """
interpolator = CloughTocherInterpolatorValues(points, (grid_x, grid_y))
for val in values:
    interpolator(val)
"""


number = 1
repeats = 1

run_ops = {
    "old": [oldclass, shared_setup],
    "subclass": [newclass_code, "\n".join([shared_setup, newclass_setup])] 
}

import timeit
for run_op in run_ops:
    data = timeit.repeat(run_ops[run_op][0], run_ops[run_op][1], number=10, repeat=10)
    print(f"# {run_op}:")
    print(data)
    print(f"# Mean: {np.mean(data)}")
    print(f"# STD: {np.std(data)}")
```

The results are:
```
# old:
[8.16984250000678, 7.453101999824867, 7.209543599979952, 7.252109300112352, 7.754185999976471, 7.364752599969506, 6.851857700152323, 6.9910080002155155, 8.941267400048673, 7.7140397999901325]
# Mean: 7.570170890027657
# STD: 0.5858832658034183
# subclass:
[0.42856690008193254, 0.43522219988517463, 0.5436153998598456, 0.5385737998876721, 0.4428130001761019, 0.40720779984258115, 0.39274090016260743, 0.43384339986369014, 0.4344679000787437, 0.4395203001331538]
# Mean: 0.4496571599971503
# STD: 0.04802008195881562
```

In this case, the operation is more than x15 faster, and the more different sets of values there are to interpolate the difference gets larger. This is very helpful for realtime applications.

I have also fixed both comments, and updated the benchmark class (in case they ever work again...)

File: scipy/interpolate/interpnd.pyx (hunk 12 above)
Comment (hovavalon) [reply scipy/interpolate/interpnd.pyx]: Fixed.

**Comment by hovavalon** [comment]
It turns out I cheated a little bit, the original code could have been made faster if only one Delaunay triangulation would have been performed for all the iterations.
Anyway, even when passing both classes the triangulations the time differences are of the same scale:
```
# old:
[5.451684499857947, 6.480922400020063, 6.2209514998830855, 6.5187896999996156, 7.7494192000012845, 7.614329699892551, 6.9236576000694185, 8.53692699992098, 8.009030600078404, 7.648563799913973]
# Mean: 7.115427599963732
# STD: 0.9009854432717213
# subclass:
[0.46789659978821874, 0.4547214999329299, 0.4393583999481052, 0.4584745999891311, 0.4434726999606937, 0.4415132000576705, 0.44243260007351637, 0.4472308000549674, 0.4540659000631422, 0.43496869993396103]
# Mean: 0.4484134999802336
# STD: 0.009615146765158272
```

**Comment by hovavalon** [comment]
Are the errors related to me?
I don't think so, but your approval of that would surely help :)

**Comment by ev-br** [comment]
Hmm... there's a slight git glitch it seems: submodule changes should not be there.
I'd do
```
$ git fetch upstream
$ git rebase upstream/main
$ git submodule update --init      # this
```

then `$ git status` should not show anything about boostmath. Then force-push

```
$ git push origin HEAD -f
```

**Comment by hovavalon** [comment]
Fixed (In a slightly different way).

**Review by ev-br — APPROVED** [review]
LGTM now!

CI failure seems unrelated.

**Comment by ev-br** [comment]
Merged. Thank you @hovavalon and congratulations with what I believe is your first scipy contribution. Keep them coming!
